unstructured/partition/base.py

```diff
@@ -0,0 +1,106 @@
+from __future__ import annotations
+
+import io
+from abc import abstractmethod
+from tempfile import SpooledTemporaryFile
+from typing import (
+    Any,
+    BinaryIO,
+    Iterator,
+    Optional,
+    Union,
+)
+
+from unstructured.documents.elements import Element
+from unstructured.partition.common import (
+    get_last_modified_date,
+    get_last_modified_date_from_file,
+)
+from unstructured.utils import lazyproperty
+
+
+class _Partitioner:
+    """Provides `.partition()` for files."""
+
+    # TODO: I think we can do better on metadata.filename. Should that only be populated when a
+    #       `metadata_filename` argument was provided to `partition_docx()`? What about when not but
+    #       we do get a `filename` arg or a `file` arg that has a `.name` attribute?
+
+    def __init__(
+        self,
+        filename: Optional[str],
+        file: Optional[Union[BinaryIO, SpooledTemporaryFile[bytes]]],
+        metadata_filename: Optional[str],
+        metadata_last_modified: Optional[str],
+    ) -> None:
+        self._filename = filename
+        self._file = file
+        self._metadata_filename = metadata_filename
+        self._metadata_last_modified = metadata_last_modified
+
+    @classmethod
+    def iter_document_elements(
+        cls,
+        filename: Optional[str] = None,
+        file: Optional[Union[BinaryIO, SpooledTemporaryFile[bytes]]] = None,
+        metadata_filename: Optional[str] = None,
+        metadata_last_modified: Optional[str] = None,
+    ) -> Iterator[Element]:
+        """Partition document into its document elements."""
+        return cls(
+            filename,
+            file,
+            metadata_filename,
+            metadata_last_modified,
+        )._iter_document_elements()
+
+    @abstractmethod
+    def _iter_document_elements(self) -> Iterator[Element]:
+        """Generate each document-element in document order."""
+        pass
```

Comment:
There are two roles that can be fulfilled by a class like this and in my experience they are usually better kept separate:

1. **Abstract interface definition.** Well handled by a `Protocol` and the `AbstractPartitioner` protocol definition above is about all you can say about these classes because they only have the one "public" method. This defines what an object needs to implement to be considered one of these things (this is _structural typing_).
2. **Base class.** A base class provides default implementations and possibly abstract-methods, although I think those can easily become heavy-handed. I'll say more about that below, but in this particular case, do we really care that folks use a method named `._iter_document_elements()` to implement the main functionality of the class? I would consider that more of a convention/template type of concern and not rigidly require that, unless you were going to inherit `iter_document_elements()` in the sub-classes, which I don't think you can (because they have different option-types).